test/server/active_tcp_listener_test.cc

```diff
@@ -0,0 +1,181 @@
+#include <memory>
+
+#include "envoy/network/filter.h"
+#include "envoy/network/listener.h"
+#include "envoy/stats/scope.h"
+
+#include "common/network/address_impl.h"
+#include "common/network/connection_balancer_impl.h"
+#include "common/network/raw_buffer_socket.h"
+#include "common/network/utility.h"
+
+#include "server/active_tcp_listener.h"
+
+#include "test/mocks/api/mocks.h"
+#include "test/mocks/common.h"
+#include "test/mocks/network/mocks.h"
+#include "test/test_common/network_utility.h"
+
+#include "gmock/gmock.h"
+#include "gtest/gtest.h"
+
+using testing::_;
+using testing::Invoke;
+using testing::NiceMock;
+using testing::Return;
+using testing::ReturnRef;
+
+namespace Envoy {
+namespace Server {
+namespace {
+
+class MockTcpConnectionHandler : public Network::TcpConnectionHandler,
```

Comment:
MockConnectionHandler virtual inheritance from ConnectionHandler is required here. The diamond is described as below.
`MockTcpConnectionHandler` <- TcpConnectionHandler <- `ConnectionHandler`
`MockTcpConnectionHandler` <- MockConnectionHandler <- `ConnectionHandler`
                            
 We want MockTcpConnectionHandler gets the mock implementation of MockConnectionHandler for free but also specialized for TcpConnectionHandler.